### PR TITLE
fix(if): nested ifs don't work properly

### LINKED ISSUE #328: Leaking (the new) if binding introduced in v1.5.2
**I'm submitting a bug report**

* **Library Version:**
1.5.2

**Please tell us about your environment:**
* **Operating System:**
OSX 10.x

* **Node Version:**
8.5.0

* **NPM Version:**
5.5.1

* **JSPM OR Webpack AND Version**
CLI

* **Browser:**
all

* **Language:**
ESNext

**Current behavior:**
https://github.com/huochunpeng/ifff

A leaking if binding behind nested if binding

Run the test app, just click next and next, all conditional DOM are not removed.

Pretty sure this is introduced by #324 which attempt to fix another issue on the new if/else implementation.

This issue doesn't exist on v1.5.1 before #324 fix.

**Expected/desired behavior:**

bring back old if.bind please...
The new if/else doesn't justify the pain...

* **What is the expected behavior?**

* **What is the motivation / use case for changing the behavior?**

### PR DESCRIPTION
**Pull request by jods4**
fixes #328, #327, #326, #322, #317, #315

### CODE REVIEW

**Comment by jods4** [comment]
This should fix all reported issues with the `if` refactoring. 

I have coded two short repro of reported bugs to confirm it's fixed and to prevent future regressions, unfortunately after more than 6 hours of trying I can't get `ComponentTester` to work in karma tests.
I would really like to commit those or something equivalent, if anyone can help.

Here are the two non-comitted tests:
```js
  describe('(nested)', () => {
    let taskQueue, component, model;

    beforeEach(() => {
      model = { a: true, b: true, clicked() {} };
      component = new ComponentTester();
      component.bootstrap(aurelia => {
        aurelia.use.standardConfiguration();
        taskQueue = aurelia.container.get(TaskQueue);
      });
      component
        .inView(`<div if.bind="a" id="a">
                   <div if.bind="b" id="b" click.delegate="clicked()">
                   </div>
                 </div>`)
        .boundTo(model);        
    });

    afterEach(() => component.dispose());

    it('should rebind even if the view is visible; issue #317', done => {
      component.create(bootstrap)
               .then(_ => {
                 let spy = spyOn(model, "clicked");
                 model.a = false;
                 taskQueue.flushMicroTaskQueue();
                 model.a = true;
                 taskQueue.flushMicroTaskQueue();
                 document.getElementById("b").click();
                 
                 expect(spy).toHaveBeenCalled();
                 done();
               });
    });

    it('should update view when rebound; issue #328', done => {
      component.create(bootstrap)
               .then(_ => {                
                 model.a = false;
                 taskQueue.flushMicroTaskQueue();
                 model.b = false;
                 model.a = true;
                 taskQueue.flushMicroTaskQueue();
 
                 expect(document.getElementById("a")).not.toBeNull();
                 expect(document.getElementById("b")).toBeNull();
                 done();
               });
    });
  });
```

File: src/if-core.js [hunk 1]
```diff
@@ -47,6 +46,12 @@ export class IfCore {
 
   _show() {
     if (this.showing) {
+      // Ensures the view is bound. 
+      // It might not be the case when the if was unbound but not detached, then rebound.
+      // Typical case where this happens is nested ifs
+      if (!this.view.isBound) {
+        this.view.bind(this.bindingContext, this.overrideContext);
+      }
```
Comment (mjwwit) [inline src/if-core.js]: More trailing spaces.

File: src/if-core.js [hunk 2]
```diff
@@ -47,6 +46,12 @@ export class IfCore {
 
   _show() {
     if (this.showing) {
+      // Ensures the view is bound.
```
Comment (mjwwit) [inline src/if-core.js]: There's a trailing space here, causing a linter error.

File: src/if.js [hunk 3]
```diff
@@ -23,6 +23,9 @@ export class If extends IfCore {
     if (this.condition) {
       this._show();
     }
+    else {
```
Comment (mjwwit) [inline src/if.js]: Same brace issue as with `src/else.js`.

File: src/else.js [hunk 4]
```diff
@@ -14,7 +14,10 @@ export class Else extends IfCore {
   bind(bindingContext, overrideContext) {
     super.bind(bindingContext, overrideContext);
     // Render on initial
-    if (!this.ifVm.condition) {
+    if (this.ifVm.condition) {
+      this._hide();
+    }
+    else {
```
Comment (mjwwit) [inline src/else.js]: According to ESLint, the `else` should be on the same line as the closing brace from the `if`.

**Comment by mjwwit** [comment]
@jods4 I've sent you a PR (jods4/templating-resources#1) with the unit-test from #330 (and additional tests), linter fixes, and actual integration tests for nested if bindings. This was a real b*tch to get working btw.

**Comment by jods4** [comment]
@mjwwit Thanks, it's merged! Although I had to disable the `ComponentTester` tests. :(

**Comment by jods4** [comment]
In addition to my own tests, we have user confirmation in #328 that the fix works for him.
@EisenbergEffect I think we can publish this?

**Comment by mjwwit** [comment]
@jods4 That sucks, I spent tons of time figuring a way around the duplicate attribute conflict... It worked for me using Karma and Chrome 61 on macOS... What issue are you running into?

**Comment by jods4** [comment]
@mjwwit I spent ton of time on that as well... which is why I gave up in the end.
I've tested in Firefox 57.0b11 on Windows 10.

A few observations I made when trying to get `ComponentTester` to work:
- It uses JSPM, whose `config.json` file locks to very old libs. When I updated everything `jspm update` I got lots of new tests failures.
- JSPM seems to -- at least sometimes if not always -- load `aurelia-templating-resources` from NPM rather than from `dist`, which is the system under test.
- `ComponentTester` only seems to work for the first test that uses it and fail for subsequent ones. So using `fit` (focused test) on any of those actually worked but never when running all of them.

**Comment by mjwwit** [comment]
Alright. I'll see if I can migrate this project to the new TypeScript build soon, maybe that'll fix these problems and finally allow for proper integration tests.

**Comment by EisenbergEffect** [comment]
@mjwwit This is a larger, more complicated, more risky project to move to TS. I'd prefer to get a few more smaller ones done first ;)

**Comment by nenadvicentic** [comment]
@EisenbergEffect Where are you tracking porting to Typescript?

**Pull request merged**